src/zulip/mod.rs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    Ok(())
+}
+
+/// Fetches the definitions of the user groups from the Team API
+fn get_user_group_definitions(
+    team_api: &TeamApi,
+    zulip_api: &ZulipApi,
+) -> Result<BTreeMap<String, Vec<usize>>, Error> {
+    let email_map = zulip_api
+        .get_users()?
+        .into_iter()
+        .map(|u| (u.email, u.user_id))
+        .collect::<BTreeMap<_, _>>();
+    let user_group_definitions = team_api
+        .get_zulip_groups()?
+        .groups
+        .into_iter()
+        .map(|(name, group)| {
+            let members = &group.members;
+            let member_ids = members
+                .iter()
+                .filter_map(|member| match member {
+                    ZulipGroupMember::Email(e) => {
+                        let id = email_map.get(e);
+                        if id.is_none() {
+                            log::warn!("no Zulip id found for '{}'", e);
+                        }
+                        id
+                    }
+                    ZulipGroupMember::Id(id) => Some(id),
+                })
+                .copied()
+                .collect::<Vec<_>>();
+            (name, member_ids)
+        })
+        .collect();
+    Ok(user_group_definitions)
+}
+
+/// Interacts with the Zulip API
+struct ZulipController {
+    /// User group name to Zulip user group id
+    user_group_ids: BTreeMap<String, ZulipUserGroup>,
+    /// The Zulip API
+    zulip_api: ZulipApi,
+    /// Whether this is a dry run or not
+    dry_run: bool,
+}
+
+impl ZulipController {
+    /// Create a new `ZulipController`
+    fn new(zulip_api: ZulipApi, dry_run: bool) -> Result<Self, Error> {
+        let user_groups = zulip_api.get_user_groups()?;
+
+        let user_group_ids = user_groups
+            .into_iter()
+            .map(|mut ug| {
+                // sort for better diagnostics
+                ug.members.sort_unstable();
+                (ug.name.clone(), ug)
+            })
+            .collect();
+
+        Ok(Self {
+            user_group_ids,
+            zulip_api,
+            dry_run,
+        })
+    }
+
+    /// Create or update a user group for the given team name and members
+    fn create_or_update_user_group(
+        &mut self,
+        user_group_name: &str,
+        member_zulip_ids: &[usize],
+    ) -> Result<(), Error> {
+        let id = self.user_group_id_from_name(&user_group_name);
+        let user_group_id = match id {
+            Some(id) => {
+                log::info!(
+                    "'{}' user group ({}) already exists on Zulip",
+                    user_group_name,
+                    id
+                );
+                id
+            }
+            None => {
+                log::info!(
+                    "no '{}' user group found on Zulip. Creating one...",
+                    user_group_name
+                );
+                self.create_user_group(
+                    &user_group_name,
+                    &format!("The {} team", user_group_name),
```

Comment:
In other synchronizations we had, the description was usually something like "Managed by the rust-lang/team repo", to tell people with admin rights not to manually change that thing.

That could also be used to identify which groups were added by the team repo, to be able to remove them once they're removed from the repo without touching manually managed groups. [Mailgun is already doing that](https://github.com/rust-lang/sync-team/blob/7af1838fdd021438e9f60b7a999d3d4b72c2853e/src/mailgun/mod.rs#L128-L139) for example.